planner/core/optimizer.go

```diff
@@ -517,62 +517,121 @@ func physicalOptimize(logic LogicalPlan, planCounter *PlanCounterTp) (plan Physi
+		pointGet      *PointGetPlan
+		batchPointGet *BatchPointGetPlan
+
+		physicalLock *PhysicalLock
+	}
+
+	tuples := make([]*Tuple, 0)
+
+	// physicalLockMap is a map for mapping PhysicalLock to PhysicalUnionScan. We need this map because when
+	// getLockWaitTime(sctx, physicalLock.Lock) returns false, we delete the relevant PhysicalUnionScan (if it exists)
+	// in the unionScanSet to avoid elimination.
+	physicalLockMap := make(map[PhysicalPlan]PhysicalPlan)
+	unionScanSet := make(map[PhysicalPlan]interface{})
+
+	iteratePhysicalPlan(p, nil, nil,
+		func(p PhysicalPlan, physicalLock *PhysicalLock, unionScan *PhysicalUnionScan) (*PhysicalLock, *PhysicalUnionScan) {
+			// PhysicalUnionScan and PhysicalLock can only relate to one point get/batch point get.
+			// So, if the plan has multiple children, we should reset physicalLockMap and unionScanSet
+			if len(p.Children()) > 1 {
+				physicalLockMap = make(map[PhysicalPlan]PhysicalPlan)
+				unionScanSet = make(map[PhysicalPlan]interface{})
+			}
+			switch x := p.(type) {
+			case *PointGetPlan:
+				tuples = append(tuples, &Tuple{
+					pointGet:     x,
+					physicalLock: physicalLock,
+				})
+				if physicalLock != nil {
+					physicalLockMap[physicalLock] = unionScan
+				}
+				if unionScan != nil {
+					unionScanSet[unionScan] = nil
+				}
+			case *BatchPointGetPlan:
+				tuples = append(tuples, &Tuple{
+					batchPointGet: x,
+					physicalLock:  physicalLock,
+				})
+				if physicalLock != nil {
+					physicalLockMap[physicalLock] = unionScan
+				}
+				if unionScan != nil {
+					unionScanSet[unionScan] = nil
+				}
+			case *PhysicalLock:
+				return x, nil
+			case *PhysicalUnionScan:
+				return nil, x
+			}
+			return nil, nil
+		})
+	if len(tuples) == 0 {
 		return p
 	}
-	if physLock == nil && unionScan == nil {
+	if len(physicalLockMap) == 0 && len(unionScanSet) == 0 {
 		return p
 	}
-	if physLock != nil {
-		lock, waitTime := getLockWaitTime(sctx, physLock.Lock)
-		if !lock {
-			return p
-		}
-		if pointGet != nil {
-			pointGet.Lock = lock
-			pointGet.LockWaitTime = waitTime
-		} else {
-			batchPointGet.Lock = lock
-			batchPointGet.LockWaitTime = waitTime
+
+	if len(physicalLockMap) != 0 {
+		for _, tuple := range tuples {
+			if tuple.physicalLock == nil {
+				continue
+			}
+
+			lock, waitTime := getLockWaitTime(sctx, tuple.physicalLock.Lock)
+			if !lock {
+				unionScan := physicalLockMap[tuple.physicalLock]
+				if unionScan != nil {
+					delete(unionScanSet, unionScan)
+				}
+				delete(physicalLockMap, tuple.physicalLock)
+				continue
+			}
+			if tuple.pointGet != nil {
+				tuple.pointGet.Lock = lock
+				tuple.pointGet.LockWaitTime = waitTime
+			} else {
+				tuple.batchPointGet.Lock = lock
+				tuple.batchPointGet.LockWaitTime = waitTime
+			}
 		}
 	}
+
 	return transformPhysicalPlan(p, func(p PhysicalPlan) PhysicalPlan {
-		if p == physLock {
+		if _, exist := physicalLockMap[p]; exist {
```

Comment:
the old code is quite tricky, only take a single chain into consideration, once they found a pointGet or batchPointGet, that must be their leaf node(datasource). So just simply sink Lock OP to the inside of *PointGet.

Once remove the limitation for multi-child. For cases of Lock -> Union(pointGet, Selection+tableReader), we still need to keep the Lock OP above it to lock those keys selected from union's right executor.